File: sdk/js/src/bincode.ts
// Minimal bincode encoder for Lichen transactions

import { PublicKey } from './publickey';
import { Instruction, Message, Transaction } from './transaction';

const textEncoder = new TextEncoder();

function encodeU64LE(value: number | bigint): Uint8Array {
  const out = new Uint8Array(8);
  const view = new DataView(out.buffer);
  view.setBigUint64(0, BigInt(value), true);
  return out;
}

function encodeU32LE(value: number): Uint8Array {
  const out = new Uint8Array(4);
  const view = new DataView(out.buffer);
  view.setUint32(0, value, true);
  return out;
}

function encodeOptionU64(value?: number): Uint8Array {
  if (value === undefined || value === null) {
    return new Uint8Array([0x00]); // None
  }
  return concat([new Uint8Array([0x01]), encodeU64LE(value)]); // Some(value)
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

function encodeBytes(data: Uint8Array): Uint8Array {
  return concat([encodeU64LE(data.length), data]);
}

function encodeString(value: string): Uint8Array {
  const encoded = textEncoder.encode(value);
  return concat([encodeU64LE(encoded.length), encoded]);
}

function encodeVec(items: Uint8Array[]): Uint8Array {
  return concat([encodeU64LE(items.length), ...items]);
}

export function hexToBytes(hex: string): Uint8Array {
  const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (clean.length % 2 !== 0) {
    throw new Error('Invalid hex string');
  }
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

function encodePubkey(pubkey: PublicKey): Uint8Array {
  const bytes = pubkey.toBytes();
  if (bytes.length !== 32) {
    throw new Error('PublicKey must be 32 bytes');
  }
  return bytes;
}

function encodeInstruction(ix: Instruction): Uint8Array {
  const programId = encodePubkey(ix.programId);
  const accounts = encodeVec(ix.accounts.map(encodePubkey));
  const data = encodeBytes(ix.data);
  return concat([programId, accounts, data]);
}

export function encodeMessage(message: Message): Uint8Array {
  const instructions = encodeVec(message.instructions.map(encodeInstruction));
  const blockhash = hexToBytes(message.recentBlockhash);
  if (blockhash.length !== 32) {
    throw new Error('Blockhash must be 32 bytes');
  }
  const computeBudget = encodeOptionU64(message.computeBudget);
  const computeUnitPrice = encodeOptionU64(message.computeUnitPrice);
  return concat([instructions, blockhash, computeBudget, computeUnitPrice]);
}

export function encodeTransaction(transaction: Transaction): Uint8Array {
  // Encode signatures as Vec<[u8; 64]> matching Rust bincode format.
  // Each signature is a hex string → 64 raw bytes. Fixed-size arrays in
  // bincode have no per-element length prefix.
  const sigBytes = transaction.signatures.map(hexSig => {
    const raw = hexToBytes(hexSig);
    if (raw.length !== 64) {
      throw new Error(`Signature must be 64 bytes, got ${raw.length}`);
    }
    return raw;
  });
  const encodedSigs = concat([encodeU64LE(sigBytes.length), ...sigBytes]);
  const messageBytes = encodeMessage(transaction.message);
  // tx_type: Native=0 (u32 LE)
  const txType = encodeU32LE(0);
  return concat([encodedSigs, messageBytes, txType]);
}

File: sdk/python/moltchain/bincode.py
"""Minimal bincode encoder for MoltChain transactions"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List

from .publickey import PublicKey


def _encode_u64(value: int) -> bytes:
    return value.to_bytes(8, byteorder="little", signed=False)


def _encode_bytes(data: bytes) -> bytes:
    return _encode_u64(len(data)) + data


def _encode_string(value: str) -> bytes:
    encoded = value.encode("utf-8")
    return _encode_u64(len(encoded)) + encoded


def _encode_vec(items: Iterable[bytes]) -> bytes:
    items_list = list(items)
    return _encode_u64(len(items_list)) + b"".join(items_list)


def _encode_pubkey(pubkey: PublicKey) -> bytes:
    raw = pubkey.to_bytes()
    if len(raw) != 32:
        raise ValueError("PublicKey must be 32 bytes")
    return raw


def _encode_hash(hex_str: str) -> bytes:
    raw = bytes.fromhex(hex_str)
    if len(raw) != 32:
        raise ValueError("Blockhash must be 32 bytes")
    return raw


@dataclass
class EncodedInstruction:
    program_id: PublicKey
    accounts: List[PublicKey]
    data: bytes


def encode_instruction(ix: EncodedInstruction) -> bytes:
    program_id = _encode_pubkey(ix.program_id)
    accounts = _encode_vec(_encode_pubkey(acc) for acc in ix.accounts)
    data = _encode_bytes(ix.data)
    return program_id + accounts + data


def encode_message(instructions: List[EncodedInstruction], recent_blockhash: str) -> bytes:
    encoded_instructions = _encode_vec(encode_instruction(ix) for ix in instructions)
    blockhash = _encode_hash(recent_blockhash)
    return encoded_instructions + blockhash


def encode_transaction(signatures: List[str], message_bytes: bytes) -> bytes:
    """Encode transaction matching Rust bincode Vec<[u8; 64]> format.

    Signatures are hex strings that map to 64 raw bytes each.
    Fixed-size arrays in bincode have no per-element length prefix.
    """
    sig_bytes = [bytes.fromhex(sig) for sig in signatures]
    for sig in sig_bytes:
        if len(sig) != 64:
            raise ValueError(f"Signature must be 64 bytes, got {len(sig)}")
    return _encode_u64(len(sig_bytes)) + b"".join(sig_bytes) + message_bytes
